kernel/timeout.c

```diff
@@ -18,6 +18,17 @@ static sys_dlist_t timeout_list = SYS_DLIST_STATIC_INIT(&timeout_list);
 
 static struct k_spinlock timeout_lock;
 
+/* On multiprocessor setups, it's possible to have multiple
+ * sys_clock_announce() calls arrive in parallel (the latest to exit
+ * the driver will generally be announcing zero ticks).  But we want
+ * the list of timeouts to be executed serially, so as not to confuse
+ * application code.  This lock wraps the announce loop, external to
+ * the nested timeout_lock.
```

Comment:
Worth pointing out that there's a reasonable worry about lock contention here: all cpus are potentially going to spin in their timer ISRs until the timeouts finish on just one of them.

The way to optimize this would be to have the secondary calls atomically increment the announce_remaining counter with their own ticks in such a way that they know the running loop will continue.  I didn't do that here for simplicity reasons (as we're trying to work around a real bug or hardware glitch in the immediate term).  But it's something we probably want.